datafusion/physical-optimizer/src/limit_pushdown_past_window.rs

```diff
@@ -0,0 +1,145 @@
+        config: &ConfigOptions,
+    ) -> datafusion_common::Result<Arc<dyn ExecutionPlan>> {
+        if !config.optimizer.enable_window_limits {
+            return Ok(original);
+        }
+        // println!("-------------------------------");
+        // println!("{}", DisplayableExecutionPlan::new(original.as_ref()).indent(false));
+        let mut latest_limit: Option<usize> = None;
+        let mut latest_min = 0;
+        let mut latest_max = 0;
+        let result = original.transform_down(|node| {
+            // println!("Transforming {:?}", node.name());
+            let mut reset = |node| -> datafusion_common::Result<
+                Transformed<Arc<dyn ExecutionPlan>>,
+            > {
+                latest_limit = None;
+                latest_min = 0;
+                latest_max = 0;
+                Ok(Transformed::no(node))
+            };
+
+            // grab the latest limit we see
+            if let Some(limit) = node.as_any().downcast_ref::<GlobalLimitExec>() {
+                latest_limit = limit.fetch();
+                latest_min = 0;
+                latest_max = 0;
+                // println!("Saving limit {:?}", latest_limit);
+                return Ok(Transformed::no(node));
+            }
+
+            // grow the limit if we hit a window function
+            if let Some(window) = node.as_any().downcast_ref::<BoundedWindowAggExec>() {
+                let Ok(expr) = window.window_expr().iter().exactly_one() else {
+                    return reset(node);
+                };
+                let frame = expr.get_window_frame();
+                if frame.units != WindowFrameUnits::Rows {
+                    return reset(node);
+                }
+                let Some(start_bound) = bound_to_i64(&frame.start_bound) else {
+                    return reset(node);
+                };
+                let Some(end_bound) = bound_to_i64(&frame.end_bound) else {
+                    return reset(node);
+                };
+                latest_min = cmp::min(start_bound, latest_min);
+                latest_min = cmp::min(end_bound, latest_min);
+                latest_max = cmp::max(start_bound, latest_max);
+                latest_max = cmp::max(end_bound, latest_max);
+
+                return Ok(Transformed::no(node));
+            }
+
+            // we can't push the limit past nodes that change row count
+            match node.cardinality_effect() {
+                CardinalityEffect::Equal => {}
+                _ => {
+                    // println!("bailing due to cardinality change");
+                    return reset(node);
+                }
+            }
+
+            // Apply the limit if we hit a sort node
+            if let Some(sort) = node.as_any().downcast_ref::<SortExec>() {
+                if let Some(fetch) = latest_limit {
+                    let window_size = (latest_max - latest_min) as usize;
+                    let sort = sort.with_fetch(Some(fetch + window_size));
```

Comment:
I am not sure if this is correct, I think it should be `fetch` + the end bound? We need at least `fetch` rows to calculate non-window expressions (following) and to get at least that amount of rows, and we need to have at least the next rows available for computing all `fetch` inputs?